### PR TITLE
a node that follows and checks

### PR DESCRIPTION
**Pull request by pro-wh**
This is proposed as the beginning of auditing tools. There's no fast sync, so you would join the network as a node, and while you catch up, you do additional processing to check invariants. Our ABCI mux system allows us to put these checks in a mux app.

cc #2277

### CODE REVIEW

**Comment by kostko** [comment]
Probably also related to #1633. cc @abukosek

**Comment by pro-wh** [comment]
new in this push:
- checking not on every block. checking on some interval. and it's implemented in this way (which nobody asked for) where it randomly chooses one height in each interval
- starting to port some sanity checks from genesis-time

**Comment by pro-wh** [comment]
from Friday and additional notes:
- mostly I've adapted genesis-time sanity checks by extracting functions that operate on abstract types. at genesis time, we call them with parts of the genesis document. at follow time, we load from the mux apps' state modules. there's some code sharing, but we'll have to be careful in the future to make sure we call these sanity checks from both places when we add new ones.
- the mux apps' state modules tend to expose methods that load entire datasets into memory, e.g. all delegation records. that's consistent with the design of the genesis-time sanity checks, which operate on an all-in-memory genesis document. for now I'm going ahead and doing things that way

**Comment by pro-wh** [comment]
new in this push and from Monday:
- fixes to increase unit test fidelity so that they pass followtool's sanity checks
- changes to sanity checks tests to make the combined system not self-contradictory
- specifying a key manager in a runtime record is now optional. key-manager-type runtimes now don't specify a (meta?) key manager. previously they would specify that they are their own key manager.
- transfers of staking tokens to/from a runtime through roothash messages now adjust the staking token's total supply so that the staking mux app's state will still add up right after the transfer. this is intended to be an interim consistent behavior before we decommission runtime transfers for #2377.

**Comment by kostko** [comment]
> specifying a key manager in a runtime record is now optional. key-manager-type runtimes now don't specify a (meta?) key manager. previously they would specify that they are their own key manager.

So this also fixes #2459?

File: go/genesis/api/api_test.go [hunk 1]
```diff
@@ -136,9 +136,9 @@ func TestGenesisSanityCheck(t *testing.T) {
 
 	testRuntimeID := hex2pk("0000000000000000000000000000000000000000000000000000000000000001")
 	testRuntime := &registry.Runtime{
-		ID:         testRuntimeID,
-		Kind:       registry.KindCompute,
-		KeyManager: testKMRuntime.ID,
+		ID:            testRuntimeID,
+		Kind:          registry.KindCompute,
+		KeyManagerOpt: &testKMRuntime.ID,
```
Comment (kostko) [inline go/genesis/api/api_test.go]: Why the rename to `KeyManagerOpt`? Whether something is optional or required shouldn't be in the name.

**Comment by pro-wh** [comment]
> So this also fixes #2459 or it just makes the key manager optional?

just makes the key manager optional

File: go/genesis/api/api_test.go (hunk 1 above)
Comment (pro-wh) [reply go/genesis/api/api_test.go]: oh I did that to make sure I would notice all usages. I'll change it back

**Comment by pro-wh** [comment]
in the latest push:
- added more corrections to the node and to the checks to make sure we always pass the checks

ready for review

File: go/keymanager/api/api.go [hunk 2]
```diff
@@ -136,9 +136,8 @@ type Genesis struct {
 	Statuses []*Status `json:"statuses,omitempty"`
 }
 
-// SanityCheck does basic sanity checking on the genesis state.
-func (g *Genesis) SanityCheck() error {
-	for _, status := range g.Statuses {
+func SanityCheckStatuses(statuses []*Status) error {
```
Comment (kostko) [inline go/keymanager/api/api.go]: Add documentation block (also all other split `Sanity*` methods).

File: go/consensus/tendermint/apps/followtool/api.go [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package followtool
+
+const (
+	AppID   uint8  = 0x98
```
Comment (kostko) [inline go/consensus/tendermint/apps/followtool/api.go]: Add documentation blocks.

File: go/keymanager/api/api.go (hunk 2 above)
Comment (pro-wh) [reply go/keymanager/api/api.go]: ok

File: go/consensus/tendermint/apps/followtool/api.go (hunk 3 above)
Comment (pro-wh) [reply go/consensus/tendermint/apps/followtool/api.go]: ok

**Comment by pro-wh** [comment]
renamed

**Comment by pro-wh** [comment]
ok all lights greenish. gonna merge

**Pull request merged**